src/module_3/evaluation_functions.py

```diff
@@ -0,0 +1,179 @@
+
+import pandas as pd
+import numpy as np
+from sklearn.model_selection import train_test_split
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import  StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import GridSearchCV
+from sklearn.compose import ColumnTransformer
+import category_encoders as ce
+from sklearn.model_selection import StratifiedKFold
+import matplotlib.pyplot as plt
+from datetime import timedelta
+from sklearn.model_selection import PredefinedSplit
+import pyarrow.parquet as pq
+from sklearn import metrics
+from sklearn.metrics import roc_auc_score, confusion_matrix
+
+
+def plot_precision_recall_vs_threshold(clf, X, y):
+    if hasattr(clf, 'predict_proba'):
+        model_probs = clf.predict_proba(X)
+        model_probs = model_probs[:, 1]
+    else:
+        model_probs = clf.decision_function(X)
+
+    model_precision, model_recall, thresholds = metrics.precision_recall_curve(y, model_probs)
+    thresholds = np.insert(thresholds, thresholds.size, 1., axis=0)
+
+    plt.figure()
+    plt.plot(thresholds, model_precision, 'r', label='Precision')
+    plt.plot(thresholds, model_recall, 'b', label='Recall')
+    plt.xlabel('Threshold Value')
+    plt.title('Precision and Recall vs Threshold')
+    plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))
+    plt.show()
+
+def threshold_classification(positive_class_probs, threshold=0.5):
+    return (positive_class_probs >= threshold).astype('int')
+
+def find_best_threshold(model, X_train, y_train, X_val, y_val, metric='fscore'):
+    if hasattr(model, 'predict_proba'):
```

Comment:
Don't Repeat Yourself. You could have a function like:
```
def predict_model_probs(model, X):
    if hasattr(model, 'predict_proba'):
        return model.predict_proba(X_train)[:, 1]
    return model.decision_function(X_train)
```
That will be reuse every time. if you need both train and val:
```
model_probs = predict_model_probs(model, X_train)
model_probs_val = predict_model_probs(model, X_val)
```